include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/grpc/status.h"
+#include "envoy/http/header_map.h"
+
+#include "google/protobuf/descriptor.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   */
+  virtual void sendMessage(const RequestType& request) PURE;
+
+  /**
+   * Finish the stream. No further operations are permitted on the stream.
+   */
+  virtual void finish() PURE;
```

Comment:
I agree, what was there before was confusing, in particular if we compare with the Finish semantics in the gRPC C++ library, which is a full close. In the model we're adopting in `AsyncClient` here, there's independent local/remote closing of streams. I will rename this to `close` to make the semantics clearer.